### PR TITLE
Minor clean / tidying up the code

### PR DESCRIPTION
**Pull request by creack**
- Scope errors when possible.
- Fix nil pointers in unmarshal.
- Align struct tags for readability.
- Explicitly discard errors with a comment on why.
- Import package order.
- Fmt verb for types.
- More comments.

### CODE REVIEW

File: coder-sdk/secrets.go [hunk 1]
```diff
@@ -19,26 +19,36 @@ type Secret struct {
...
 // SecretWithValueByName gets the Coder secret with its value by its name.
 func (c *Client) SecretWithValueByName(ctx context.Context, name, userID string) (*Secret, error) {
+	// Lookup the secret from the name.
 	s, err := c.SecretByName(ctx, name, userID)
 	if err != nil {
 		return nil, err
 	}
+	// Pull the secret value.
+	// NOTE: This is racy, but acceptable. If the secret is gone or the permission changed since we looked up the id,
+	//       the call will simply fail and surface the error to the user.
```
Comment (cmoog) [inline coder-sdk/secrets.go:37]: I don't see this as particularly actionable... if the secret is gone then the CLI will surface a 404 as expected. Changes in permissions would also result in the expected error.

File: internal/cmd/auth.go [hunk 2]
```diff
@@ -3,42 +3,41 @@ package cmd
...
 	"go.coder.com/flog"
 )
 
-// requireAuth exits the process with a nonzero exit code if the user is not authenticated to make requests
+// requireAuth exits the process with a nonzero exit code if the user is not authenticated to make requests.
 func requireAuth() *coder.Client {
 	client, err := newClient()
 	if err != nil {
-		flog.Fatal("%v", err)
+		flog.Fatal("%s", err)
```
Comment (cmoog) [inline internal/cmd/auth.go:18]: I've been meaning to remove this in favor of returning a proper error.

File: internal/cmd/shell.go [hunk 3]
```diff
@@ -196,9 +203,10 @@ func heartbeat(ctx context.Context, c *websocket.Conn, interval time.Duration) {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			err := c.Ping(ctx)
-			if err != nil {
-				flog.Fatal("\nFailed to ping websocket: %v, exiting...", err)
+			if err := conn.Ping(ctx); err != nil {
+				// TODO: Don't fatal, it would skip the TTY state restore and result in a broken terminal.
```
Comment (cmoog) [inline internal/cmd/shell.go:207]: Ah! Nice catch.

**Review by cmoog — APPROVED** [review]
Thanks 😄

File: internal/x/xterminal/terminal_windows.go [hunk 4]
```diff
@@ -85,29 +88,38 @@ func (s ResizeEvent) equal(s2 *ResizeEvent) bool {
...
+		var lastEvent *ResizeEvent
 		for {
 			select {
 			case <-ctx.Done():
 				return
 			case <-ticker.C:
-				width, height, err := terminal.GetSize(int(windows.Handle(termfd)))
+				width, height, err := terminal.GetSize(int(windows.Handle(termFD)))
 				if err != nil {
+					// TODO: Consider doing a debug log and continue instead of sopping the routine.
```
Comment (coadler) [inline internal/x/xterminal/terminal_windows.go:110]:
```suggestion
					// TODO: Consider doing a debug log and continue instead of stopping the routine.
```

**Review by coadler — APPROVED** [review]
Just some random typos I found. Great improvements :smiley:

File: ci/integration/setup_test.go [hunk 5]
```diff
@@ -32,7 +32,7 @@ func init() {
 
 // build the coder-cli binary and move to the integration testing bin directory
 func build(path string) error {
-	tar := fmt.Sprintf("coder-cli-linux-amd64.tar.gz")
```
[the remark this answers is not in the document]
Comment (cmoog) [reply ci/integration/setup_test.go:35]: Yeah, that one is particularly embarrassing... leftover from when it was MacOS and Linux. 😭

File: cmd/coder/main.go [hunk 6]
```diff
@@ -15,31 +15,35 @@ import (
...
+// Using a global for the version so it can be set at build time using ldflags.
+var version string = "unknown"
 
 func main() {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	// If requested, spin up the pprof webserver.
 	if os.Getenv("PPROF") != "" {
 		go func() {
```
Comment (tychoish) [inline cmd/coder/main.go:27]: I wish we put panic handlers in all go routines.

File: cmd/coder/main.go [hunk 7]
```diff
@@ -15,31 +15,35 @@ import (
...
 			log.Println(http.ListenAndServe("localhost:6060", nil))
 		}()
 	}
 
+	// TODO: Document why this is needed. Only does something on windows.
```
Comment (tychoish) [inline cmd/coder/main.go:32]: I'd be inclined to avoid adding TODOs to the code base? I think it's better to either do the thing if its small or open issues?

File: cmd/coder/main.go [hunk 8]
```diff
@@ -15,31 +15,35 @@ import (
...
+		_ = xterminal.Restore(os.Stdout.Fd(), stdoutState)
+	}()
 
 	app := cmd.Make()
 	app.Version = fmt.Sprintf("%s %s %s/%s", version, runtime.Version(), runtime.GOOS, runtime.GOARCH)
 
-	err = app.ExecuteContext(ctx)
-	if err != nil {
+	if err := app.ExecuteContext(ctx); err != nil {
+		// NOTE: The cmd lib will handle the logs, no need to do it here as well, simply exit with error code.
```
Comment (tychoish) [inline cmd/coder/main.go:46]: what about this error? is it interesting?

File: coder-sdk/client.go [hunk 9]
```diff
@@ -36,5 +32,7 @@ func (c *Client) http() (*http.Client, error) {
 			Secure:   c.BaseURL.Scheme == "https",
 		},
 	})
+	// TODO: Consider setting a timeout.
+	// TODO: Consider having a persistent client to allow for single connection.
```
Comment (tychoish) [inline coder-sdk/client.go:36]: add bug reports for this? I think timeouts are great. 

I'd also be inclined in doing something to pool/reuse clients (e.g. https://github.com/deciduosity/utility/blob/main/http.go)

File: coder-sdk/env.go [hunk 10]
```diff
@@ -81,50 +84,39 @@ type CreateEnvironmentRequest struct {
...
-	var env *Environment
-	err := c.requestBody(
-		ctx,
-		http.MethodPost, "/api/orgs/"+orgID+"/environments",
-		req,
-		env,
-	)
-	return env, err
+	var env Environment
+	if err := c.requestBody(ctx, http.MethodPost, "/api/orgs/"+orgID+"/environments", req, &env); err != nil {
```
Comment (tychoish) [inline coder-sdk/env.go:88]: sprintf rather than string concatenation?

File: coder-sdk/error.go [hunk 11]
```diff
@@ -2,18 +2,20 @@ package coder
...
 	"net/http"
 	"net/http/httputil"
 
 	"golang.org/x/xerrors"
 )
 
-// ErrNotFound describes an error case in which the requested resource could not be found
-var ErrNotFound = xerrors.Errorf("resource not found")
+// ErrNotFound describes an error case in which the requested resource could not be found.
+var ErrNotFound = errors.New("resource not found")
```
Comment (tychoish) [inline coder-sdk/error.go:13]: I thought we were pro-xerrors?

File: coder-sdk/env.go [hunk 12]
```diff
@@ -81,50 +84,39 @@ type CreateEnvironmentRequest struct {
...
 	var envs []Environment
-	err := c.requestBody(
-		ctx,
-		http.MethodGet, "/api/orgs/"+orgID+"/members/"+userID+"/environments",
-		nil,
-		&envs,
-	)
-	return envs, err
+	if err := c.requestBody(ctx, http.MethodGet, "/api/orgs/"+orgID+"/members/"+userID+"/environments", nil, &envs); err != nil {
+		return nil, err
```
Comment (tychoish) [inline coder-sdk/env.go:98]: should we be wrapping these as discussed?

File: coder-sdk/request.go [hunk 13]
```diff
@@ -4,51 +4,62 @@ import (
...
-	req, err := http.NewRequestWithContext(ctx, method, c.BaseURL.String()+path, bytes.NewReader(body))
+
+	// Create the http request.
+	req, err := http.NewRequestWithContext(ctx, method, c.BaseURL.String()+path, payload)
 	if err != nil {
 		return nil, xerrors.Errorf("create request: %w", err)
 	}
+	// TODO: Consider setting content-type / encoding / accept headers.
+
+	// Execute the request.
```
Comment (tychoish) [inline coder-sdk/request.go:38]: I don't think it's necessary to have comments to describe function calls.

also, omit the todo or opening a ticket?

File: internal/activity/pusher.go [hunk 14]
```diff
@@ -4,13 +4,14 @@ import (
...
-	"cdr.dev/coder-cli/coder-sdk"
 	"golang.org/x/time/rate"
 
+	"cdr.dev/coder-cli/coder-sdk"
+
 	"go.coder.com/flog"
 )
 
-const pushInterval = time.Minute
+const pushInterval = 1 * time.Minute
```
Comment (tychoish) [inline internal/activity/pusher.go:14]: I wouldn't do this.

File: internal/cmd/cmd.go [hunk 15]
```diff
@@ -92,13 +91,13 @@ $ coder completion fish > ~/.config/fish/completions/coder.fish
 	Run: func(cmd *cobra.Command, args []string) {
 		switch args[0] {
 		case "bash":
-			cmd.Root().GenBashCompletion(os.Stdout)
+			_ = cmd.Root().GenBashCompletion(os.Stdout) // Best effort.
```
Comment (tychoish) [inline internal/cmd/cmd.go:94]: I would only make the best effort comment once?

File: internal/cmd/shell.go [hunk 16]
```diff
@@ -95,32 +96,39 @@ func sendResizeEvents(ctx context.Context, termfd uintptr, process wsep.Process)
...
 	if err != nil {
-		return err
+		return xerrors.Errorf("find environment: %w", err)
 	}
 
-	termfd := os.Stdout.Fd()
+	termFD := os.Stdout.Fd()
 
-	tty := terminal.IsTerminal(int(termfd))
+	// Check if the client is running an interactive session.
```
Comment (tychoish) [inline internal/cmd/shell.go:109]: this seems redundant to the code. 

I'm generally hesitant on adding comments that are redundant to the code in cases like this where the comments just describe what the code does, which just can get out of date.

File: internal/cmd/shell.go [hunk 17]
```diff
@@ -196,9 +203,10 @@ func heartbeat(ctx context.Context, c *websocket.Conn, interval time.Duration) {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			err := c.Ping(ctx)
-			if err != nil {
-				flog.Fatal("\nFailed to ping websocket: %v, exiting...", err)
+			if err := conn.Ping(ctx); err != nil {
+				// TODO: Don't fatal, it would skip the TTY state restore and result in a broken terminal.
+				// NOTE: Prefix with \r\n to attempt to have clearer output as we might still be in raw mode.
```
Comment (tychoish) [inline internal/cmd/shell.go:208]: can the prefix be a constant that we can reuse for these kinds of situations?

File: internal/sync/sync.go [hunk 18]
```diff
@@ -91,74 +91,70 @@ func (s Sync) syncPaths(delete bool, local, remote string) error {
...
+	go func() { _, _ = io.Copy(os.Stderr, process.Stderr()) }() // Best effort.
 
-	err = process.Wait()
-	if code, ok := err.(wsep.ExitError); ok {
-		return fmt.Errorf("%s exit status: %v", prog, code)
-	}
-	if err != nil {
+	if err = process.Wait(); err != nil {
+		if code, ok := err.(wsep.ExitError); ok {
+			return fmt.Errorf("%s exit status: %d", prog, code)
```
Comment (tychoish) [inline internal/sync/sync.go:112]: we should be consistent, I think about which Errorf questions.

File: internal/sync/sync.go [hunk 19]
```diff
@@ -91,74 +91,70 @@ func (s Sync) syncPaths(delete bool, local, remote string) error {
...
 		}
-
 		return err
 	}
 	return nil
 }
 
 func (s Sync) handleDelete(localPath string) error {
-	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```
Comment (tychoish) [inline internal/sync/sync.go:157]: I wouldn't multiply by 1?

File: internal/sync/sync.go [hunk 20]
```diff
@@ -215,7 +211,7 @@ var ErrRestartSync = errors.New("the sync exited because it was overloaded, rest
 
 // workEventGroup converges a group of events to prevent duplicate work.
 func (s Sync) workEventGroup(evs []timedEvent) {
-	cache := make(eventCache)
+	cache := make(eventCache, len(evs))
```
Comment (tychoish) [inline internal/sync/sync.go:214]: what Add do? this seems suspicious?

File: internal/x/xterminal/terminal_windows.go [hunk 21]
```diff
@@ -85,29 +88,38 @@ func (s ResizeEvent) equal(s2 *ResizeEvent) bool {
 // ResizeEvents sends terminal resize events when the dimensions change.
 // Windows does not have a unix.SIGWINCH equivalent, so we poll the terminal size
 // at a fixed interval
-func ResizeEvents(ctx context.Context, termfd uintptr) chan ResizeEvent {
-	events := make(chan ResizeEvent)
-	ticker := time.NewTicker(time.Millisecond * 100)
+func ResizeEvents(ctx context.Context, termFD uintptr) chan ResizeEvent {
+	events := make(chan ResizeEvent, 1) // Use a buffered chan to avoid blocking.
```
Comment (tychoish) [inline internal/x/xterminal/terminal_windows.go:92]: isn't blocking a symptom of a bug? if it works without buffering, I would leave it buffered?

**Review by tychoish — CHANGES_REQUESTED** [review]

File: coder-sdk/secrets.go (hunk 1 above)
Comment (cmoog) [reply coder-sdk/secrets.go:37]: Ah, I see. Thanks 👍 .

File: internal/cmd/auth.go (hunk 2 above)
Comment (creack) [reply internal/cmd/auth.go:18]: https://github.com/cdr/coder-cli/issues/114

File: internal/x/xterminal/terminal_windows.go (hunk 21 above)
Comment (creack) [reply internal/x/xterminal/terminal_windows.go:92]: I'll update the comment to add more context about the reasoning.

We want to send an initial event to signal the current size of the window. I was planning to do this in a different PR as this one ended up being mostly about styling/cleaning.

At a higher level, the idea is that when dealing with the channel, the runtime notifies us via a non-blocking chan, so if we are consuming it on an unbuffered chan, we lose the event.
In this specific case, we are dealing with windows and it is not actually a signal, but it maps directly with how it works on unix.

File: internal/sync/sync.go (hunk 20 above)
Comment (creack) [reply internal/sync/sync.go:214]: We are creating a map where we know exactly how big it is going to be in advance, so we can pre-allocate the memory for it.

File: internal/sync/sync.go (hunk 19 above)
Comment (creack) [reply internal/sync/sync.go:157]: for consistency and readability, I think it is best to always specify the full time, it shows it is intended and not an oversight. As it is a const, this is handled at compile time and have no impact and the runtime, it is purely aesthetic.

With this in mind, if you think it is best to remove it, we sure can.

File: internal/sync/sync.go (hunk 18 above)
Comment (creack) [reply internal/sync/sync.go:112]: I was wondering about it, as it was already there, I opted not to change it as there might have been a reason?

File: internal/cmd/shell.go (hunk 17 above)
Comment (creack) [reply internal/cmd/shell.go:208]: This is a one-off that should be avoided, as we should control how/what we output. Promoting this to a const would imply it is supported and might encourage it to be used and for just two chars, especially CRLF, I am not sure it would add value.

File: internal/cmd/shell.go (hunk 16 above)
Comment (creack) [reply internal/cmd/shell.go:109]: I am not sure most people make the relative between IsTerminal and interactive. We are dealing with somewhat advanced low level programming, following the execution flow in plain engligh helps understanding.

Out of date comments is indeed a problem, but it is on us not to let that happen :)

File: coder-sdk/env.go (hunk 10 above)
Comment (creack) [reply coder-sdk/env.go:88]: There is a mix of both a bit everywhere in the repo. Here I just stopped the error without changing how the string is created.

Created #115 to track this.

File: cmd/coder/main.go (hunk 8 above)
Comment (creack) [reply cmd/coder/main.go:46]: I'll reword the comment to be clearer. That error is already checked and log in the cmd lib. It is indeed an interesting one :)

File: cmd/coder/main.go (hunk 7 above)
Comment (creack) [reply cmd/coder/main.go:32]: Will remove the todos in favor of issues.

File: cmd/coder/main.go (hunk 6 above)
Comment (creack) [reply cmd/coder/main.go:27]: Hmm, panics are developer errors, not really expected in production. As a safety net on a service to log/alert about a bug makes sense, and in general, for anything expected to be long-running, I do agree, we should have panic handler, but here, we are in a CLI and specifically in a debug handler, so it is probably not necessary.

File: coder-sdk/error.go (hunk 11 above)
Comment (creack) [reply coder-sdk/error.go:13]: From chatting with @nhooyr we are using xerrors for the >go1.13 formatting/warps but also for the stack trace. Here we don't have any wrap/formatting, and as we are dealing with a "sentinel" error, having a trace indicated error.gol13 doesn't provide value and can lead to confusion.

File: coder-sdk/env.go (hunk 12 above)
Comment (creack) [reply coder-sdk/env.go:98]: When there is only one error path, imho it is not necessary, but more is better, I'll add it.

File: internal/cmd/cmd.go (hunk 15 above)
Comment (creack) [reply internal/cmd/cmd.go:94]: I actually strongly disagree, it is about the concept, there is an error, it is discarded, we indicate it is on purpose, this should be done each and every time. Ideally with more details, but at the very least with a message showing the intent.

File: internal/activity/pusher.go (hunk 14 above)
Comment (creack) [reply internal/activity/pusher.go:14]: Can you extend on why? imho, it makes things easier to read and shows that it is a meant like the actual interval and not a unit

File: coder-sdk/request.go (hunk 13 above)
Comment (creack) [reply coder-sdk/request.go:38]: I am not sure everyone makes the relation between `req.Do` and effectively executing the request, especially when having `http.NewRequest` earlier. Having comments to explain the flow in plain English makes things easier to follow and maintain.

If you think it is just too much, I"ll remove it.

File: coder-sdk/client.go (hunk 9 above)
Comment (creack) [reply coder-sdk/client.go:36]: Will do.

Using a 3rd party pooling here might be overkill, but it would be interesting in our services

File: coder-sdk/error.go (hunk 11 above)
Comment (nhooyr) [reply coder-sdk/error.go:13]: I think it's good to be consistent though. If we use `xerrors.New``, I think we're good.

File: coder-sdk/client.go (hunk 9 above)
Comment (tychoish) [reply coder-sdk/client.go]: The pool seems to only be needed when you're playing with multiple transports, I think it's fine as is for now.

File: coder-sdk/request.go (hunk 13 above)
Comment (tychoish) [reply coder-sdk/request.go:38]: I'm still for removing. 

Comments are just one more thing that can get crufty, and while I agree that the interface is intuitive, I would hate that we should sprinkle descriptions of other package's APIs throughout calling code.

File: internal/activity/pusher.go (hunk 14 above)
Comment (tychoish) [reply internal/activity/pusher.go:14]: I always assume that people typo'd `10*time.Minute` here, and think that time constants are clear enough.

File: internal/activity/pusher.go [hunk 22]
```diff
@@ -19,27 +20,28 @@ type Pusher struct {
 	source string
 
 	client *coder.Client
-	rate   *rate.Limiter
+	rate   *rate.Limiter // Use a rate limiter to control the sampling rate.
```
Comment (tychoish) [inline internal/activity/pusher.go:23]: I would omit this comment too.

File: internal/cmd/cmd.go (hunk 15 above)
Comment (tychoish) [reply internal/cmd/cmd.go:94]: I think putting a higher level comment that says "we ignore errors when generating completion because they're not critical and the tool degrades gracefully" is very clear, whereas "best effort" doesn't really capture that decision or background as clearly (particularly when redundant.) 

I think the ignored error itself does a pretty good job of communicating intent, though I agree that some kind of comment is useful.

I'm not strongly committed to this view.

File: internal/cmd/shell.go (hunk 16 above)
Comment (tychoish) [reply internal/cmd/shell.go:109]: The best way to avoid having out of date comments is to avoid writing code that needs comments :) 

Perhaps the `tty` variable would be more clear if it was named `isInteractive` or something similar. It could also be combined onto the next line. 

I think given how internal/low level (I suppose), I think it's fine

File: internal/cmd/shell.go (hunk 17 above)
Comment (tychoish) [reply internal/cmd/shell.go:207]: I think putting "LineFeed" as constant gives it a name which makes the comment less required?

I feel like adding `\r\n` to the ends of something is never just a one-off, but I'm far less familiar with this code base.

File: internal/sync/sync.go (hunk 18 above)
Comment (tychoish) [reply internal/sync/sync.go:112]: I'm not sure, I can't imagine a way that it would matter, and it feels like this is the right kind of PR to make that change. I don't feel particularly strongly at all.

File: internal/sync/sync.go (hunk 19 above)
Comment (tychoish) [reply internal/sync/sync.go:157]: I'm still for removing, in a "less is more" sort of approach.

File: internal/sync/sync.go (hunk 20 above)
Comment (tychoish) [reply internal/sync/sync.go:214]: Interesting, I guess not seeing what type `eventCache` is makes this harder. I'm all down for preallocation, but we're far enough from the context that it was hard to read/guess. 

Frankly, it'd make more sense if there was a constructor that did all of this, rather than just having it at the beginning of a worker pool.

**Review by tychoish — APPROVED** [review]
I don't think anything is blocking merging this, but we can talk a little bit more about any of the details. I think most of these changes are an improvement.
